toolchain/sem_ir/type_iterator.h

```diff
@@ -0,0 +1,185 @@
+      -> std::variant<SemIR::TypeId, SymbolicType>;
+
+  // Get the instructions in the specific's instruction block as an ArrayRef.
+  auto GetSpecificArgs(SemIR::SpecificId specific_id)
+      -> llvm::ArrayRef<SemIR::InstId>;
+
+  // Push all arguments from the array into the work queue.
+  auto PushArgs(llvm::ArrayRef<SemIR::InstId> args) -> void;
+
+  // Push an instruction's type value into the work queue, or a marker if the
+  // instruction has a symbolic value.
+  auto PushInstId(SemIR::InstId inst_id) -> void;
+
+  // Push the next step into the work queue.
+  auto Push(WorkItem item) -> void;
+
+  File* sem_ir_;
+  llvm::SmallVector<WorkItem> work_list_;
+};
+
+class TypeIterator::Step {
+ public:
+  // Followed by generic parameters.
+  struct ClassStart {
+    SemIR::ClassId class_id;
+    SemIR::TypeId type_id;
+  };
+  // Followed by its fields.
+  struct StructStart {
+    SemIR::TypeId type_id;
+  };
+  // Followed by its members.
+  struct TupleStart {
+    SemIR::TypeId type_id;
+  };
+  // Followed by generic parameters.
+  struct InterfaceStart {
+    SemIR::InterfaceId interface_id;
+  };
+  // Followed by the bit width.
+  struct IntStart {
+    SemIR::TypeId type_id;
+  };
+  // Followed by the type and bound.
+  struct ArrayStart {
+    SemIR::TypeId type_id;
+  };
+  // Followed by the pointee type.
+  struct PointerStart {};
+
+  // A type value.
+  struct ConcreteType {
+    SemIR::TypeId type_id;
+  };
+  // A symbolic type value, constrained by `facet_type_id`.
+  struct SymbolicType {
+    // Either a FacetType or the TypeType singleton.
+    SemIR::TypeId facet_type_id;
+  };
+  // A symbolic template type value.
+  struct TemplateType {};
+  // A non-type value, which can be found as a generic parameter for a type.
+  struct Value {
+    // TODO: Include the type of the value somewhere.
+  };
+
+  // A *Start type, but without anything following it, so we also skip the
+  // matching `End`.
+  struct StartOnly {
+    using Any =
+        std::variant<ClassStart, StructStart, TupleStart, InterfaceStart>;
+    Any any;
+  };
+  // A *Start type, with nested steps after it, and then a matching `End`.
+  struct StartWithEnd {
+    using Any =
+        std::variant<ClassStart, StructStart, TupleStart, InterfaceStart,
+                     ArrayStart, IntStart, PointerStart>;
+    Any any;
+  };
+
+  // Closes the scope of a `StartWithEnd`.
+  struct End {};
+
+  // Iteration is complete.
+  struct Done {};
+
+  // Each step is one of these.
+  using Any = std::variant<ConcreteType, SymbolicType, TemplateType, Value,
+                           StartOnly, StartWithEnd, End, Done>;
+
+  template <class T>
+  auto Is() const -> bool {
+    return std::holds_alternative<T>(any);
+  }
+
+  // Converts from any individual step.
+  template <class T>
+    requires std::constructible_from<Any, T>
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  Step(T any) : any(any) {}
```

Comment:
Could this be:

```suggestion
  Step(Any any) : any(std::move(any)) {}
```

(note, I at least think the explicit move would be recommended by clang-tidy)

Reply:
The move is warned _against_ by clang-tidy since the type is trivially movable.

This does not work though:
```
toolchain/sem_ir/type_iterator.cpp:156:10: error: no viable conversion from returned value of type 'Step::Done' to function return type 'Step'
  156 |   return Step::Done();
      |          ^~~~~~~~~~~~
./toolchain/sem_ir/type_iterator.h:176:3: note: candidate constructor not viable: no known conversion from 'Step::Done' to 'Any' (aka 'variant<ConcreteType, SymbolicType, TemplateType, Value, StartOnly, StartWithEnd, End, Done>') for 1st argument
  176 |   Step(Any any) : any(any) {}
      |   ^    ~~~~~~~
```

Looking at the [constructors of variant](https://en.cppreference.com/w/cpp/utility/variant/variant), you'd think the converting constructor (4) would work here, it's the same constructor we're calling to initialize the member, and the constructor is not marked `explicit`. Any yet.

We're already implicitly converting to construct Step, so maybe it won't also implicitly convert to construct an argument to Step? That's not the rule I am familiar with re: limiting implicit conversions but...

This all happens cuz we have a struct-with-variant-inside so there's two levels of structure going on, instead of just a single tagged-union type though.

I added a comment.